Fail loudly with corrupted Parquet statistics

zhenxiao opened the pull request:
@nezihyigitbasi @dain @findepi @qqibrow

zhenxiao commented:
thank you, @electrum 
I get comments addressed. will do following task to add dataSourceID

nezihyigitbasi commented:
In addition to knowing which file is bad, it would be easier to debug if we know which column's stats is corrupted. By looking at the stack trace of the exception we can see for which type we have corrupted stats, but if we have many columns of that type then that's not useful. In short, an exception message like: `Corrupted statistics for column %s in Parquet file %s` would be pretty useful.

nezihyigitbasi commented:
There is also a risk of breaking user queries that scan Parquet files with corrupted stats. Do we want to have a flag to revert to the current behavior just in case (which we can eventually remove and make this new behavior the default behavior)?

findepi commented:
> Do we want to have a flag to revert to the current behavior just in case (which we can eventually remove and make this new behavior the default behavior)?

@nezihyigitbasi IMO we should fail on corrupted stats by default -- unless some widely used Parquet writing program is notorious for writing corrupted stats. 
And yes, we should have safety flag to disable the check, at least for some transition period.

zhenxiao commented:
@nezihyigitbasi @findepi @electrum thanks ur comments
get comments addressed, added dataSourceId for Parquet, added configuration to control whether fail query when scan Parquet file with corrupted statistics. During transition period, disable the check, we could enable it, and probably remove the config, to fail whenever hitting corrupted statistics

nezihyigitbasi commented:
> IMO we should fail on corrupted stats by default -- unless some widely used Parquet writing program is notorious for writing corrupted stats.

@findepi My comment was confusing probably, sorry. I totally agree with you. We should fail on corrupt stats by default. @zhenxiao looks like this PR sets the flag to false by default btw.

nezihyigitbasi commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java:
```diff
@@ -903,6 +904,19 @@ public HiveClientConfig setUseParquetColumnNames(boolean useParquetColumnNames)
+    }
+
+    @Config("hive.parquet.fail-on-corrupted-statistics")
+    @ConfigDescription("Fail query when scan Parquet file with corrupted statistics")
```
`Fail when scanning Parquet files with corrupted statistics`

nezihyigitbasi commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveSessionProperties.java:
```diff
@@ -225,6 +226,11 @@ public HiveSessionProperties(HiveClientConfig hiveClientConfig, OrcFileWriterCon
                         false),
+                booleanProperty(
+                        PARQUET_FAIL_WITH_CORRUPTED_STATISTICS,
+                        "Parquet: Fail query when scan Parquet file with corrupted statistics",
```
"Fail when scanning Parquet files with corrupted statistics"

nezihyigitbasi commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/HdfsParquetDataSource.java:
```diff
@@ -89,15 +97,15 @@ private void readInternal(long position, byte[] buffer, int bufferOffset, int bu
         }
         catch (Exception e) {
-            throw new PrestoException(HIVE_FILESYSTEM_ERROR, format("Error reading from %s at position %s", name, position), e);
+            throw new PrestoException(HIVE_FILESYSTEM_ERROR, format("Error reading from %s at position %s", id.toString(), position), e);
```
Don't need the `toString` call.

nezihyigitbasi commented on presto-parquet/src/main/java/com/facebook/presto/parquet/predicate/Predicate.java:
```diff
@@ -41,8 +42,10 @@ public boolean matches(Map<ColumnDescriptor, DictionaryDescriptor> dictionaries)
      * Statistics to determine if a column is only null
      * @param statistics column statistics
+     * @param id Parquet file name
+     * @param failOnCorruptedParquetStatistics whether to fail query when scan Parquet file with corrupted statistics
```
`when scanning a Parquet file`

nezihyigitbasi commented on presto-parquet/src/main/java/com/facebook/presto/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -149,16 +151,20 @@ else if ((type.equals(BIGINT) || type.equals(TINYINT) || type.equals(SMALLINT) |
                 IntStatistics intStatistics = (IntStatistics) statistics;
-                // ignore corrupted statistics
                 if (intStatistics.genericGetMin() > intStatistics.genericGetMax()) {
+                    if (failOnCorruptedParquetStatistics) {
```
You can create a method like this one (feel free to find a better name) to simplify & get rid of repetition:

```
private static void failWithCorruptionException(boolean failOnCorruptedParquetStatistics, String column, ParquetDataSourceId id, Object min, Object max)
{
    if (failOnCorruptedParquetStatistics) {
        throw new ParquetCorruptionException(format("Corrupted statistics for column %s in Parquet file %s: min %s, max %s", column, id, min, max));
    }
}
```

Then calls will look like:

```
failWithCorruptionException(failOnCorruptedParquetStatistics, column, id, intStatistics.genericGetMin(), intStatistics.genericGetMax());
```

zhenxiao commented:
thank you @nezihyigitbasi 
I get comments addressed, squash into one commit, added testcases for ignore and fail on corrupted statistics

nezihyigitbasi commented on presto-parquet/src/main/java/com/facebook/presto/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -280,6 +282,13 @@ else if (isVarcharType(type) && columnDescriptor.getType() == PrimitiveTypeName.
+    private static void failWithCorruptionException(boolean failOnCorruptedParquetStatistics, String column, ParquetDataSourceId id, Statistics statistics)
+    {
+        if (failOnCorruptedParquetStatistics) {
+            throw new ParquetCorruptionException(format("Corrupted statistics for column %s in Parquet file %s: %s", column, id, statistics));
```
I hope all children of `Statistics` type have `toString` implemented.

nezihyigitbasi replied:
One minor change we can make to the error message is to add quotes/brackets to set the boundaries of the column/file name and stats information:

```
Corrupted statistics for column "column_name" in Parquet file "file_name": [stats_string]
```

Currently it looks like the following, which is a little bit hard to follow:

```
Corrupted statistics for column DateColumn in Parquet file testFile: min: 200, max: 100, num_nulls: 0
```

zhenxiao commented on presto-parquet/src/main/java/com/facebook/presto/parquet/ParquetCorruptionException.java:
```diff
@@ -13,12 +13,10 @@
 import static java.lang.String.format;
 
 public class ParquetCorruptionException
-        extends IOException
```
OK, will keep it

zhenxiao commented:
thank you, @nezihyigitbasi 
comments addressed, added testcases for all types

nezihyigitbasi commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java:
```diff
@@ -205,6 +211,9 @@ public static ParquetPageSource createParquetPageSource(
             if (e instanceof PrestoException) {
                 throw (PrestoException) e;
             }
+            if (nullToEmpty(e.getMessage()).trim().startsWith("Corrupted statistics for column")) {
```
Instead of relying on the error message, can't we just do:

```
if (e instanceof ParquetCorruptionException)
    throw new PrestoException(HIVE_BAD_DATA, e);
}
```

zhenxiao commented:
thank you, @nezihyigitbasi 
get comments addressed
@dain @electrum @findepi your comments are welcome

nezihyigitbasi commented:
Thanks @zhenxiao, merged.
